Posix and portability improvements

jpco opened the pull request:
These changes improve _es_' ability to build and function correctly across different Unices.  At first I made these changes:

 - Use the OS-provided exit macros (`WIFEXITED`, `WIFSIGNALED`, etc.) if they exist, accounting for the fact that `WCOREDUMP` is still not universally adopted.  This makes _es_ build on Haiku.
 - ~~Just use `setpgid()` always, instead of the old way of `setsid()` if available, `setpgid()` otherwise, and `setpgrp()` yet otherwise.  This makes _es_ behave normally in terms of process groups and sessions (e.g., `ps` output includes backgrounded processes.)~~ Removed due to a number of knock-on effects.  Will be followed up on in a later PR.
 - Use `waitpid()` instead of `wait3()`.  This fixes #125, fixes an autoconf whine, and also allows management of `Proc *`s to be simplified a heck of a lot.

I added test cases to verify these changes.  Then I tested on Linux (which is what I just run on my machine), OmniOS, OpenIndiana, macOS, FreeBSD, Haiku, and Debian Hurd.  These revealed a couple other portability problems which I fixed:

- The `SOLARIS`-related macros, ironically, broke building on Solaris-derived OSes.  Because the code they're guarding doesn't make much sense these days (based on my testing), I removed those macros entirely.
- In prim-sys.c: Switched from `ioctl(TIOCSPGRP, ...)` to `tcsetpgrp()`.  The latter is specified in POSIX and available in all these OSes; the former is an implementation detail of a couple OSes and unavailable in others.
- In access.c: Got rid of `NGROUPS`, which is not always available.  Instead, in POSIX, the value `NGROUPS` was used for can be retrieved at runtime from the `getgroups` function.

I also resolved all but one of the remaining autoconf whines, because they were pretty mechanical.  Fixing `AC_HEADER_STDC` involves litigating the header files included in _es_, which I didn't want to get into at the moment, since at this point I got things working pretty much correctly on all OSes I'd tested.

A couple smaller problems I didn't resolve:

 - ~~On several OSes (Haiku, FreeBSD, macOS, Hurd) I couldn't get the `+core` part of the exit status to work.  This is the case for the changed wait macros I use here, as well as the original ones that have been in _es_ forever.  I feel like there's something very fundamental I'm missing there.  This impacts the newly-added `tests/wait.es`.~~ Fixed except for Haiku, which is weird enough that I'm not worried about it.
 - ~~In several OSes (FreeBSD, the Solarises, macOS), `printf` has a documented bug where `printf %b 'foo\0bar'` doesn't work right, which impacts `tests/trip.es`.~~ Fixed.

memreflect commented:
`$&time` is broken by this change because `getrusage(RUSAGE_CHILDREN, ...)` gets the cumulative resource usage for all waited-for child processes, not just the most recently waited-for process (that's what made `waitN()` so ridiculously convenient, except on systems like HP-UX that don't fill in the `struct rusage*` parameter, hence `AC_FUNC_WAIT3`).
To fix this, you just need to do the following:

1. Declare `static struct rusage ru_save` and `struct rusage ru_new` inside `dowait()`.
2. Call `getrusage(RUSAGE_CHILDREN, &ru_new)`.
3. If the result of `getrusage()` is not -1 and `rusage != NULL`, store the result of `ru_new.ru_stime - ru_save.ru_stime` in `rusage->ru_stime` and the same for the `.ru_utime` members.
4. Set `ru_save = ru_new`, so the next call to `dowait()` will be able to correctly calculate the time used.

Some platforms provide a [timersub()](https://linux.die.net/man/3/timersub) macro to deal with subtraction of two `struct timeval*`, but it's not standard, so either a default implementation would be needed, or an Autoconf test to disable `BUILTIN_TIME` would be needed.

The entire [&lt;sys/resource.h&gt;](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/sys/resource.h.html) header ([until POSIX.1-2024](https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/sys_resource.h.html)) and [getrusage()](https://pubs.opengroup.org/onlinepubs/009695399/functions/getrusage.html) are also optional outside of XSI systems:

> **POSIX.1-2001, SUSv3**
> The standard defines two levels of conformance: _POSIX conformance_, which is a baseline set of interfaces required of a conforming system; and _XSI Conformance_, which additionally mandates a set of interfaces (the "XSI extension") which are only optional for POSIX conformance.
> &mdash; [standards(7)](https://linux.die.net/man/7/standards)

This means the variant of `$&time` relying upon the less precise `times()` function needs to be retained unless XSI systems are the requirement instead of POSIX systems.

There's more that you could do like switching exclusively to `sigaction()` and `sigsetjmp()`, or using `pid_t` and `ssize_t` instead of assuming they fit in an `int` and `long`, but this is a good first step in the right direction.

jpco commented:
Thank you for pointing out the XSI thing, that's my bad -- will fix.  Shame I can't yank out all that code just yet :)

As far as the `$&time` stuff goes -- I am and always have been confused about the precise semantics of `getrusage()`.  If what you're saying is true, then when I run this (on _es_ including this PR):
```
; ./es
; sleep 1
; sleep 1
; time sleep 1
     1r     0.0u     0.0s	sleep 1
; time sleep 1
     1r     0.0u     0.0s	sleep 1
```
shouldn't I be seeing `3r` and `4r` in the output of those two `time` commands?

memreflect commented:
> shouldn't I be seeing `3r` and `4r` in the output of those two `time` commands?

The "real" time has nothing to do with `getrusage()`, so that number doesn't matter.
What i'm talking about are the "user" time and the "system" time (`0.0u` and `0.0s`), which are obtained using `getrusage()`:

```
; {for (_ = $path) {time seq 1000000 > /dev/null}} | nl -w 2
 1           1r     0.1u     0.0s       seq 1000000
 2           0r     0.3u     0.0s       seq 1000000
 3           0r     0.5u     0.0s       seq 1000000
 4           0r     0.7u     0.0s       seq 1000000
 5           0r     0.9u     0.0s       seq 1000000
 6           1r     1.1u     0.0s       seq 1000000
 7           0r     1.3u     0.0s       seq 1000000
 8           0r     1.5u     0.0s       seq 1000000
 9           0r     1.6u     0.0s       seq 1000000
10           0r     1.8u     0.0s       seq 1000000
11           1r     2.0u     0.0s       seq 1000000
12           0r     2.2u     0.0s       seq 1000000
13           0r     2.4u     0.0s       seq 1000000
```

Notice how the user time continues increasing?
Contrast that with the output of the `time` utility on FreeBSD for the same code:

```
; local (fn-time = ()) {for (_ = $path) {time seq 1000000 > /dev/null}} |[2] nl -w 2
 1              0.18 real         0.18 user         0.00 sys
 2              0.18 real         0.18 user         0.00 sys
 3              0.18 real         0.18 user         0.00 sys
 4              0.18 real         0.18 user         0.00 sys
 5              0.18 real         0.18 user         0.00 sys
 6              0.18 real         0.18 user         0.00 sys
 7              0.18 real         0.18 user         0.00 sys
 8              0.18 real         0.18 user         0.00 sys
 9              0.18 real         0.18 user         0.00 sys
10              0.18 real         0.18 user         0.00 sys
11              0.18 real         0.18 user         0.00 sys
12              0.19 real         0.19 user         0.00 sys
13              0.18 real         0.18 user         0.00 sys
```

I hope that clarifies what i meant. :)

> Thank you for pointing out the XSI thing, that's my bad -- will fix. Shame I can't yank out all that code just yet :)

Agreed.
Still, it does work, though i'm not sure why the `times()` code forks twice, unless it makes measuring things easier?

> I am and always have been confused about the precise semantics of `getrusage()`.

Yes, it is rather odd, but as far as timing things is concerned, it is much the same as using `times()`.
In both cases, the times for child processes are cumulative, so you need to subtract the previous total time from the current total time to determine the amount of time taken for a single process.

Btw, you can drop the definition of the `SA_INTERRUPT` flag as well since you're removing SunOS cruft.
[sigaction(3V) on SunOS 4.1.3](https://man.freebsd.org/cgi/man.cgi?query=sigaction&apropos=0&sektion=0&manpath=SunOS+4.1.3&arch=default&format=html) without `SA_INTERRUPT` automatically restarted system calls like `wait()` when the associated signal handler returned, similar to the `SA_RESTART` flag usable on XSI systems.
With `SA_INTERRUPT`, you'd instead see `wait()` return -1 with `errno==EINTR`, which is the default behavior of the POSIX `sigaction()` function.

jpco commented:
> The "real" time has nothing to do with getrusage(), so that number doesn't matter.

Well... that's embarrassing that I didn't even check where the getrusage() output was being used.  Thanks for clarifying.  Fixed.

I've now also switched from the `S_IFREG`/`S_IFDIR`/etc. macros in access.c to the POSIX-specified `S_ISREG`/etc. macros, which was the last bit of code that didn't work with `-D_POSIX_C_SOURCE=200112L`.  (You have to manually define `NSIG` if compiling in strict POSIX.1-2001 mode, but that's fine for at least what I care about at the moment.)

You suggest multiple good follow-through improvements here, but I'm not terribly interested in pursuing them at the moment.  Obviously, I'm not exactly gifted at reasoning about C portability and standards compliance, so unless I'm making some kind of measurable, concrete improvement, I'd prefer to do as little hacking about as necessary.

jpco commented:
Added test cases so I don't miss a goof with globbing in the future. :)

It turns out commit 574358b resolves a longstanding oddness I've seen where my window manager doesn't kill backgrounded processes on exit!  More practical wins for standard-respecting behavior :+1:

memreflect commented:
> * On several OSes (Haiku, FreeBSD, macOS, Hurd) I couldn't get the `+core` part of the exit status to work.

It turns out FreeBSD's `sleep` utility is sandboxed by its "Capsicum" stuff, which prevents `sleep` from generating a core file for some reason i don't fully understand.
I'm not sure why/how the `cap_enter()` syscall prevents core file generation or why a `sleep` utility even needs sandboxed.
Weirdly, using something like `while {sleep 1}` or `fork {sleep 1}` works, unlike similar constructs in other shells.
A command that doesn't use Capsicum, such as `seq 100000000000 > /dev/null` also works.

Given the fact that `sleep 1 &` followed by `kill -QUIT $! ; wait $!` behaves mostly the same across the sh implementations i tested, i consider it to be a system incompatibility like `printf`, so you would need to replace the command with something else &mdash; a different system utility or something added to the _es_ source code to make the test work reliably.

jpco commented:
Aha, good catch.  It didn't occur to me this would be something specific about certain provided binaries rather than a global "are dumps on?" state.  I guess this is due to capsicum preventing sandboxed binaries from accessing the file system?

Anyway, at this point I'm thinking it would be easiest (for both this and the `printf` thing) to just bundle our own necessary files.  I verified that a handwritten super-simple `sleep` binary did properly dump core on FreeBSD.  _Rc_ does some of this bundling already (see their `tripping.c`), but I'd want something a little "appendable" like with our multi-file `test/tests` directory.

jpco commented:
Actually, trying to ensure we "give back" the terminal on exit is kind of a pain (unless we use `atexit()` and that scares me).  I think it's good enough to just take the terminal after waiting.

memreflect commented:
> Actually, trying to ensure we "give back" the terminal on exit is kind of a pain (unless we use atexit() and that scares me). I think it's good enough to just take the terminal after waiting.

Yes, it's good enough if _es_ is executed either as a login shell or from a shell that restores the process group after waiting for _es_ to exit/terminate as _es_ does with commit 2fb1141 to mimic a job control shell.

But when those conditions are not met, problems still occur.
For example, `es -c newpgrp` still causes the "poof" issue when job control is disabled in bash.
I think `fn newpgrp {fork {$&newpgrp ; $*}}` in `initial.es` fixes things so long as someone doesn't override that definition, but it defeats the purpose that `$&newpgrp` was meant to solve &mdash; making the shell itself a process group leader.

Another possibility to restore the terminal process group on exit is to invent yet another new primitive that is executed by `%interactive-loop` and `$&batchloop` to restore the process group ID and the terminal group to that of its parent process when the shell is about to exit.

[Here's a sample implementation of the idea.](https://github.com/memreflect/es-shell/tree/jpco-posix)
I didn't add the new primitive to the man page because i just wanted to get something working and needed a name that sort of fit.

memreflect commented on proc.c:
```diff
@@ -69,21 +52,63 @@ extern int efork(Boolean parent, Boolean background) {
+
+extern void newpgrp(void) {
+	setpgid(0, 0);
+	espgid = getpgid(0);
```
All usages of [`getpgid(0)` (XSI)](https://pubs.opengroup.org/onlinepubs/009695399/functions/getpgid.html) need replaced with [`getpgrp()` (POSIX)](https://pubs.opengroup.org/onlinepubs/009695399/functions/getpgrp.html) unless you're targeting [POSIX.1-2008](https://pubs.opengroup.org/onlinepubs/9699919799.2008edition/functions/getpgid.html).

jpco replied:
Thanks, fixed.

jpco commented:
> But when those conditions are not met, problems still occur.
> For example, `es -c newpgrp` still causes the "poof" issue when job control is disabled in bash.

Yeah, okay, it's worth fixing.

In my mind, the major challenge can be seen with `grep 'exit(.*)' *.c`.  _Es_ exits abnormally in several cases and it would be rather painful to turn those crash+error messages into crash+error+"poof" events.

> Another possibility to restore the terminal process group on exit is to invent yet another new primitive that is executed by %interactive-loop and $&batchloop to restore the process group ID and the terminal group to that of its parent process when the shell is about to exit.

I've gone back and forth on whether the "restore the terminal pgrp" behavior should be explicitly via a primitive or implicitly on exit.  As much as I'm always suggesting moving stuff "out into the shell", I also generally think the shell runtime should handle cleaning up behind the user (this is part of what I dislike about the job control design in the `jobcontrol` branch -- the `$&procfree` primitive is absolutely the kind of thing the shell runtime should be doing itself.)

My thinking right now is that restoring the terminal is in the category of cleanup the shell runtime should do, like reaping processes and managing memory.

I think the way to do it would be to create wrapper functions around `exit()` for some calls to switch to `cleanexit()` (for terminal-restoring exits) and some to do `directexit()` for the "unconditionally exit the child shell" cases.  Longer-term I would like to move as many of these `exit()`s as possible to exceptions, but given some of these are for `malloc` errors, I'd rather do a quicker improvement in the meantime.

memreflect commented:
> > But when those conditions are not met, problems still occur.
> > For example, `es -c newpgrp` still causes the "poof" issue when job control is disabled in bash.
> 
> Yeah, okay, it's worth fixing.

It is, but that can be left for a different PR if necessary.
My example was built on this PR purely because the code necessary was mostly already there.

> My thinking right now is that restoring the terminal is in the category of cleanup the shell runtime should do, like reaping processes and managing memory.

I think the same thing, but i wanted to try doing it with a primitive first to see if it was even possible.

> I think the way to do it would be to create wrapper functions around `exit()` for some calls to switch to `cleanexit()` (for terminal-restoring exits) and some to do `directexit()` for the "unconditionally exit the child shell" cases. Longer-term I would like to move as many of these `exit()`s as possible to exceptions, but given some of these are for `malloc` errors, I'd rather do a quicker improvement in the meantime.

That might be a good idea.
POSIX specifies `_exit()`, but that doesn't do everything `exit()` does.
In particular, `_exit()` can lead to data loss because open streams aren't flushed, so your proposal is more reliable on that front.

jpco commented:
Man, this `setpgid` change is really more of a pain than I expected it would be.  Feels like I'm implementing half of job control just to keep things sane as a non-job-control shell.  It also feels a little funny that _es_ is now doing more in terms of job control than Bash does with `set +m`... Maybe all this new `tc.*pgrp` stuff be behind the `JOB_PROTECT` define?

The change has also necessitated calling `$&newpgrp` in a script for the first time (yay?).  I have one script I use fairly often that does some setup and then forks a graphical program in the background.  If I don't `$&newpgrp` in that script, then the graphical program exits when my terminal does.  This is due to commit [574358b](https://github.com/wryun/es-shell/pull/128/commits/574358b1f0990fabb80b8905763129cb6edb579f) -- when called non-interactively, a backgrounded process isn't put into a new process group, and (I believe) the terminal kills the foreground process group on exit.

An additional wrinkle is that, if I source my script like `. script`, then it runs in the current shell process, which is already the session leader, so `$&newpgrp` has no effect (and my backgrounded process dies when the terminal exits.)  So to work around this, in the script, instead of just invoking `program &`, I need to do 
```
fork {
  $&newpgrp
  program &
}
```
which is a lot of working-around requiring a fair bit of knowledge of both how _es_ does things and terminal/pgrp magic.  Ugh.  It would be nicer if we could do the `$&newpgrp {program &}` thing I propose in #84 but that feels like it's getting into _actually_ implementing job control.

Also, using `setpgid` for backgrounded processes as-is makes it so that they sit around defunct after they exit:
```
; sleep 0 &
38657
; ps -o pid,ppid,pgid,tty,command
    PID    PPID    PGID TT       COMMAND
  24559   24557   24559 pts/1    /usr/local/bin/es
  38657   24559   38657 pts/1    [sleep] <defunct>
  38659   24559   24559 pts/1    /usr/bin/ps -o pid,ppid,pgid,tty,command
```
I don't know off the top of my head what happens with the `setsid()` behavior -- perhaps the process is orphaned and `init` takes over reaping it.  A job control shell is typically set up to do a WNOHANG wait before each prompt to clean these up, but maybe we're fine just letting old processes sit around, I dunno.  As with the `$&newpgrp {program &}` thing, I have a proposal for this (adding a `-n` flag to `$&wait` to do a WNOHANG wait) in #84, but again, I don't want to get into implementing (the rest of) job control with this PR.

memreflect commented:
Maybe only allow `$&newpgrp` to execute in child processes, i.e. `if (hasforked)` ?
It would avoid further searches for ways to reset the process group ID on shell exit at least.

While that defeats the intended purpose of `$&newpgrp` (see the comments above `%prompt` in `esrc.haahr`), i can't really think of any more convenient solution other than the following, which is just the same idea expressed in _es_ code instead:

```
fn-newpgrp = $&noreturn @ * {
  fork {
    $&newpgrp
    $*
  }
}
```

And if either of those options are attractive, then maybe `$&fork` and `$&background` should just accept a flag like `-g` to call `setpgid()` in the child process, and `$&newpgrp` could disappear entirely.

I think all of those ideas are mostly the same concept as `$&newpgrp` accepting arguments anyway, which you have suggested in the past.
The only difference is that no arguments to the new `fn-newpgrp`, or `$&fork -g`, would just do needless work to set the process group ID and then exit; no arguments to `$&background -g` would just set the process group ID without affecting the terminal process group and turn into a zombie.

jpco commented:
Okay, I'm rolling back the entire `setpgrp` pile of changes here in order to propose a new PR devoted just to them.  It feels like this PR has grown into "the setpgrp stuff, and the other stuff", and, especially since the "setpgrp stuff" is turning out to be pretty behavior-changing and needs consideration by itself, I'd rather break it up into separate PRs for more hygienic pull requests.

jpco commented:
Alright, verified the current commit builds and passes `make test` on FreeBSD, Omni OS, and the Hurd.

memreflect commented on stdenv.h:
```diff
@@ -301,17 +296,32 @@ extern int getgroups(int, int *);
+#endif
 
+#ifndef 	WIFEXITED
+# define	WIFEXITED(status)	(!SIFSIGNALED(status))
```
`SIFSIGNALED -> WIFSIGNALED`

Sorry i didn't catch it earlier.

jpco replied:
I'm just glad you caught it now!

memreflect commented:
Overall, these changes look good to me.

And i didn't comment on it earlier, but i noticed that switching to `waitpid()` allowed you to greatly simplify `proc.c`, too.
It was a bit confusing to follow when i initially looked at the old code, so i really like your improvements there. :)
